File: dataset/output/MISTRAL_codestral-latest/Task187_MISTRAL_codestral-latest.java
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.ReentrantLock;

public class H2O {
    private Semaphore hydrogen;
    private Semaphore oxygen;
    private ReentrantLock lock;

    public H2O() {
        hydrogen = new Semaphore(2);
        oxygen = new Semaphore(0);
        lock = new ReentrantLock();
    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
        hydrogen.acquire();
        lock.lock();
        releaseHydrogen.run();
        oxygen.release();
        lock.unlock();
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        oxygen.acquire(2);
        lock.lock();
        releaseOxygen.run();
        hydrogen.release(2);
        lock.unlock();
    }
}

File: dataset/output/MISTRAL_codestral-latest/Task187_MISTRAL_codestral-latest.cpp
#include <semaphore.h>

class H2O {
private:
    sem_t hSem;
    sem_t oSem;

public:
    H2O() {
        sem_init(&hSem, 0, 2);
        sem_init(&oSem, 0, 0);
    }

    void hydrogen(function<void()> releaseHydrogen) {
        sem_wait(&hSem);
        releaseHydrogen();
        sem_post(&oSem);
    }

    void oxygen(function<void()> releaseOxygen) {
        sem_wait(&oSem);
        sem_wait(&oSem);
        releaseOxygen();
        sem_post(&hSem);
        sem_post(&hSem);
    }
};